lib/ui/pages/barcode.dart

```diff
@@ -0,0 +1,162 @@
+class _QRCode extends StatelessWidget {
+  final SeedResponse seedResponse;
+
+  const _QRCode(this.seedResponse);
+
+  @override
+  Widget build(BuildContext context) {
+    return Column(
+      mainAxisAlignment: MainAxisAlignment.center,
+      crossAxisAlignment: CrossAxisAlignment.center,
+      children: [
+        QrImage(
+          data: seedResponse.seed,
+        ),
+        Text(
+          seedResponse.seed,
+          textAlign: TextAlign.center,
+        ),
+        _QRCodeTimer(seedResponse.expiration)
+      ],
+    );
+  }
+}
+
+/// The _QRCodeTimer class shows a count down timer for the quantity of seconds
+/// remaining for the QR Code's validity. A Stateless Widget is used to keep
+/// track of the expiration and previous time so that while counting down, the
+/// widget can determine if needs to refresh the widget any longer.
+class _QRCodeTimer extends StatefulWidget {
+  final DateTime timeout;
+
+  const _QRCodeTimer(this.timeout);
+
+  @override
+  _QRCodeTimerState createState() => _QRCodeTimerState();
+}
+
+class _QRCodeTimerState extends State<_QRCodeTimer> {
+  DateTime prevTime;
+
+  /// On the onset, we run setTimer to star the timer process
+  _QRCodeTimerState() : prevTime = DateTime.now() {
+    setTimer(const Duration(seconds: 1));
+  }
+
+  /// isExpired tells the user if the timer has expired.
+  bool get isExpired {
+    var now = DateTime.now();
+    return now.isAfter(widget.timeout);
+  }
+
+  /// timeLeft displays a string that either gives the expiration in seconds or
+  /// says "Expired" when the code expires.
+  String get timeLeft {
+    if (isExpired) {
+      return 'Expired';
+    }
+
+    var now = DateTime.now();
+    var dif = widget.timeout.difference(now);
+    return 'Expiration: ${dif.inSeconds}';
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    var style = const TextStyle(
+      fontSize: 20,
+    );
+
+    // We maintain consistency with font size, but change the color depending on
+    // whether the code is expired or not.
+    if (isExpired) {
+      style = style.copyWith(color: Colors.red);
+    }
+
+    return Text(
+      timeLeft,
+      style: style,
+    );
+  }
+
+  /// Sets a timer that will wait 1 second, then use setState to change the
+  /// prevTimer value. The prevTimer value is used to determine just how long
+  /// has passed since the last time this function was called. Setting state
+  /// will cause the widget to re-render and display the new time.
+  resetTimer() {
+    final now = DateTime.now();
+    // Step 1: Determine if we have exceeded the timeout period.
+    if (now.isAfter(widget.timeout)) {
+      // Do nothing
+      return;
+    }
+
+    // Step 2: just set the timer to 1 second later.
+    setTimer(const Duration(seconds: 1));
+  }
+
+  /// setTimer just creates a new Timer to run reset timer again.
+  setTimer(Duration duration) {
+    Timer(duration, () {
+      setState(() {
+        prevTime = DateTime.now();
+      });
```

Comment:
**question:** can you elaborate a little more on your approach here and why you decided to store the previous time this was called? I see you mention what is is used for in the docstring, but I am looking for a little more rationale behind why you chose this approach (keeping track of `prevTime`) altogether.

Reply:
Yes, the ORIGINAL intent of this function was to call a timer almost exactly 1 second. Dart's timer function isn't precise (much like JavaScript's). Meaning that If you set a timer with a duration of 1 second, it could take 1001 or more milliseconds to activate.

In the past, using a periodic (or JS interval) has bitten me in the butt, so I would usually do some quick math to determine how much time has ACTUALLY passed and set the timer for 1 second, minus any overage from the previous timeout function. However, I eventually scraped that as I had 'bigger fish to fry'.

This approach is more suitable for functions that have much tighter timing tolerances (less than 1 second).

prevTimer is kind of an artifact of this approach. However, prevTimer also provides a means to re-render the Widget by using setState, so it worked out in the end.

Reply:
Thanks for the context and explanation. I am still wondering why we need to store or keep track of `prevTime`. We are trying to show the time left that the QR code is valid right? Couldn't we remove this `prevTime` altogether? The logic that actually displays the time left to the user actually calculates the difference between whatever time it's triggered to rebuild and the expiration of the QR code, it doesn't use `prevTime` right? There may be a different way to update that widget to rebuild instead of throwing in a dummy variable, could you think of any?

Reply:
We don't need to store prevTime, specifically. I was initially under the impression that one of the Widget's member variables needed to change before the Widget would re-render. However, the StatefulWidget documentation implies that calling `setState(() {})` (i.e. no values within the closure) would schedule a rebuild without having to change anything.

Reading online suggests that there are ways to force Widgets to redraw, but most of these means seem to be anti-patterns that I won't worry about touching.

However, to be totally honest, I rarely use StatefulWidgets in applications. I prefer to create ChangeNotifiers to hold data and use either Consumer objects (sparingly) and Selector objects (mostly).